client.go

```diff
@@ -619,22 +619,29 @@ func createJoinMessages(joinedChannels map[string]bool, channels ...string) ([]s
 
 	sb := strings.Builder{}
 	sb.WriteString(baseMessage)
+	channelsWritten := 0
 
 	for _, channel := range channels {
 		channel = strings.ToLower(channel)
 		// If the channel already exists in the map we don't need to re-join it
-		if joinedChannels[channel] {
+		c.channelsMtx.Lock()
+		if c.channels[channel] {
+			c.channelsMtx.Unlock()
 			continue
 		}
-		if sb.Len()+len(channel)+2 > maxMessageLength {
+		c.channelsMtx.Unlock()
+		if sb.Len()+len(channel)+2 > maxMessageLength || (!c.rateLimiter.isUnlimited() && channelsWritten >= c.rateLimiter.joinLimit) {
 			joinMessages = append(joinMessages, sb.String())
 			sb.Reset()
 			sb.WriteString(baseMessage)
+			channelsWritten = 0
 		}
 		if sb.Len() == len(baseMessage) {
```

Comment:
Nitpick: Now that we store the amount of channels written into this stringbuilder, we don't have to be fancy with this length comparison
```suggestion
		if channelsWritten == 0 {
```